### PR TITLE
[Merged by Bors] - feat(RingTheory/Derivation/Basic): add lemmas about the derivation of a fraction

### PR DESCRIPTION
**Pull request by Command-Master**
Add lemmas about the derivation of a fraction in a field.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by j-loreaux** [comment]
Seems straightforward to me.

bors merge

**Comment by Command-Master** [comment]
Sorry about that, I haven't noticed your comment before adding `leibniz_zpow`.

File: Mathlib/RingTheory/Derivation/Basic.lean [hunk 1]
```diff
@@ -439,6 +439,40 @@ theorem leibniz_inv {K : Type*} [Field K] [Module K M] [Algebra R K] (D : Deriva
   · exact D.leibniz_of_mul_eq_one (inv_mul_cancel ha)
 #align derivation.leibniz_inv Derivation.leibniz_inv
```
Comment (mattrobball) [inline Mathlib/RingTheory/Derivation/Basic.lean:441]:
```suggestion

section DerivationOfFraction

variable {K : Type*} [Field K] [Module K M] [Algebra R K] (D : Derivation R K M) 
```

File: Mathlib/RingTheory/Derivation/Basic.lean [hunk 2]
```diff
@@ -439,6 +439,40 @@ theorem leibniz_inv {K : Type*} [Field K] [Module K M] [Algebra R K] (D : Deriva
...
 
+theorem leibniz_div {K : Type*} [Field K] [Module K M] [Algebra R K]
```
Comment (mattrobball) [inline Mathlib/RingTheory/Derivation/Basic.lean:442]:
```suggestion
theorem leibniz_div (a b : K) : D (a / b) = b⁻¹ ^ 2 • (b • D a - a • D b) := by
```

File: Mathlib/RingTheory/Derivation/Basic.lean [hunk 3]
```diff
@@ -439,6 +439,40 @@ theorem leibniz_inv {K : Type*} [Field K] [Module K M] [Algebra R K] (D : Deriva
...
+    (D : Derivation R K M) (a b : K) : D (a / b) = b⁻¹ ^ 2 • (b • D a - a • D b) := by
+  simp only [div_eq_mul_inv, leibniz, leibniz_inv, inv_pow, neg_smul, smul_neg, smul_smul, add_comm,
+    sub_eq_add_neg, smul_add]
+  rw [← inv_mul_mul_self b⁻¹, inv_inv]
+  ring_nf
+
+theorem leibniz_div_const {K : Type*} [Field K] [Module K M] [Algebra R K]
```
Comment (mattrobball) [inline Mathlib/RingTheory/Derivation/Basic.lean:449]:
```suggestion
theorem leibniz_div_const (a b : K) (h : D b = 0) : D (a / b) = b⁻¹ • D a := by
```

File: Mathlib/RingTheory/Derivation/Basic.lean [hunk 4]
```diff
@@ -439,6 +439,40 @@ theorem leibniz_inv {K : Type*} [Field K] [Module K M] [Algebra R K] (D : Deriva
...
+    (D : Derivation R K M) (a b : K) (h : D b = 0) : D (a / b) = b⁻¹ • D a := by
+  simp only [leibniz_div, inv_pow, h, smul_zero, sub_zero, smul_smul]
+  rw [← mul_self_mul_inv b⁻¹, inv_inv]
+  ring_nf
+
+lemma leibniz_zpow {K : Type*} [Field K] [Module K M] [Algebra R K]
```
Comment (mattrobball) [inline Mathlib/RingTheory/Derivation/Basic.lean:455]:
```suggestion
lemma leibniz_zpow {K : Type*} (a : K) (n : ℤ) : D (a ^ n) = n • a ^ (n - 1) • D a := by
```

File: Mathlib/RingTheory/Derivation/Basic.lean [hunk 5]
```diff
@@ -439,6 +439,40 @@ theorem leibniz_inv {K : Type*} [Field K] [Module K M] [Algebra R K] (D : Deriva
...
+    omega
+  · rw [h, zpow_neg, zpow_natCast, leibniz_inv, leibniz_pow, inv_pow, ← pow_mul, ← zpow_natCast,
+      ← zpow_natCast, nsmul_eq_smul_cast K, zsmul_eq_smul_cast K, smul_smul, smul_smul, smul_smul]
+    trans (-n.natAbs * (a ^ ((n.natAbs - 1 : ℕ) : ℤ) / (a ^ ((n.natAbs * 2 : ℕ) : ℤ)))) • D a
+    · ring_nf
+    rw [← zpow_sub₀ ha]
+    congr 3
+    · norm_cast
+    omega
+
```
Comment (mattrobball) [inline Mathlib/RingTheory/Derivation/Basic.lean:475]:
```suggestion

end DerivationOfFraction
```

File: Mathlib/RingTheory/Derivation/Basic.lean (hunk 1 above)
Comment (mattrobball) [reply Mathlib/RingTheory/Derivation/Basic.lean:441]: It is always a balance when to make the variables global vs local in scope. I think a `section` here is appropriate to improve readability. The variables are still declared very close by in the file.

Comment (Command-Master) [reply Mathlib/RingTheory/Derivation/Basic.lean:441]: Alright, I'll also insert `leibniz_inv` to it, which also uses those same variables

**Comment by mattrobball** [comment]
bors merge

**Pull request closed without merging**